### PR TITLE
Fail with a better error message on invalid variable structure in `Module.apply()`

### PR DESCRIPTION
**Pull request by mohitreddy1996**
#489 - if the collection of variables of a kind is immutable and the variable doesn't exist in the collection, we cannot initialize it (even if the RNGs are provided).

### CODE REVIEW

**Comment by avital** [comment]
Can you please add a test?

Also I'd like a second review from @jheek to see if I didn't miss any edge case.

**Comment by avital** [comment]
Wow, I had PR comments from months ago that were unsubmitted :( Apologies @mohitreddy1996 -- do you think you want to get this PR updated and merged? Alternatively, if you're busy maybe best to close the PR in the meanwhile

**Comment by jheek** [comment]
@mohitreddy1996 -- this PR has a merge conflict now. Could you take a look?

File: flax/core/scope.py [hunk 1]
```diff
@@ -297,7 +297,15 @@ def variable(self, col: str, name: str, init_fn: Callable[..., T],
                *init_args) -> Variable[T]:
     """Create a Variable."""
     self.reserve(name)
-    if not self.has_variable(col, name):
+    collection = self.collection(col=col, mutable=True)
+
+    # If the collection is immutable and the variable does not exist
+    # in it, we cannot initialize it.
+    if isinstance(collection, FrozenDict) and (name not in collection):
+      raise ValueError(f"Kind: {col} is immutable and "
```
Comment (marcvanzee) [inline flax/core/scope.py:305]: Could you use a custom error message instead? https://github.com/google/flax/blob/master/flax/errors.py

**Review by marcvanzee — CHANGES_REQUESTED** [review]

**Comment by avital** [comment]
This PR is inactive, and the original issue (#489) has been resolved. So I'm closing it.
